RFCs/0012-Test-Adapter-Lookup.md

```diff
@@ -0,0 +1,91 @@
+# 0012 Test Adapter Lookup
+
+## Summary
+This note outlines the proposed changes for lookup and initialization of the test adapters.
+
+## Motivation
+Following is a list of issues with current design
+* Test discovery fails for a solution containing projects referring to multiple versions of the same adapter.
+* Upgrade of test adapter is not honored as the adapters get cached in the host process.
+* Diagnostic information related to which adapter was used for which source and what adapters are loaded in not available.
+
+## Principles
+1. The adapter referenced in the project should be used for test execution
+2. Performance should not degrade.
+3. Adapter lookup logic is consistent across runs.
+4. Adapter lookup logic should be clear and consistent for both IDE/Editor and CLI runs.
+
+## Adapter Sources
+Here is the list of sources TestPlatform looks up for test adapters.
+
+| Source                   | Remarks                                                                        |
+|--------------------------|--------------------------------------------------------------------------------|
+| TestAdapterPath          | Adapter is specified by user in the runsettings file or via cli arguments      |
+| Nuget                    | Adapter ships as a nuget package, doesn't copy to output directory             |
+| Project output directory | Adapter ships as a nuget package, copies to output directory                   |
+| Extensions directory     | Adapters packaged with VS, available in the Extensions directory               |
+| VSIX                     | Adapter ships as a vsix.                                                       |
+
+## Proposed Changes
+Here are the proposed changes based on the sources available for lookup of the adapters.
+
+### Priority order for Lookup
+Following is the priority order for the lookup of the adapters
+
+| Priority | Source                   |
+|----------|--------------------------|
+| 1        | Nuget                    |
+| 2        | TestAdapterPath          |
+| 3        | Project output directory |
+| 4        | Extensions directory     |
+| 5        | VSIX                     |
+
+### Details for lookup
+
+#### Nuget
+This lookup is based on the nuget package references. This works only for IDE and not for CLI, as this information is only available in IDE.
```

Comment:
`dotnet test` can figure this information out though. Is that going to be similar to dll based entry points(`vstest.console.exe` and `dotnet vstest`) as well?